### PR TITLE
[Tests-Only][full-ci] Update phpstan to latest version

### PR DESCRIPTION
**Pull request by dpakach**
Related - owncloud/QA#642

Update phpstan to 0.12

### CODE REVIEW

**Comment by phil-davis** [comment]
https://drone.owncloud.com/owncloud/user_ldap/2436/2/4
`[ERROR] Found 277 errors`
:(

**Comment by dpakach** [comment]
Still fails with these errors
```
 ------ -------------------------------------------------------------------------------------------------------------------------------- 
  Line   lib/Access.php                                                                                                                  
 ------ -------------------------------------------------------------------------------------------------------------------------------- 
  654    Result of && is always false.                                                                                                   
  654    Result of || is always false.                                                                                                   
  654    Strict comparison using === between class-string<OC\Group\Backend> and 'OCA\\User_LDAP\\Group…' will always evaluate to false.  
  654    Strict comparison using === between class-string<OC\Group\Backend> and 'OCA\\User_LDAP\\Group…' will always evaluate to false.  
  972    Parameter #1 $link of method OCA\User_LDAP\ILDAPWrapper::search() expects resource, array given.                                
  972    Parameter #2 $baseDN of method OCA\User_LDAP\ILDAPWrapper::search() expects string, array<string> given.                        
 ------ -------------------------------------------------------------------------------------------------------------------------------- 

 ------ --------------------------------------- 
  Line   lib/Wizard.php                         
 ------ --------------------------------------- 
  1070   Variable $state might not be defined.  
 ------ ---------------------------------------

**Comment by phil-davis** [comment]
I raised PR #552 to just bump phpstan to 0.12 but not actually enable it yet.

There is too much change needed here, and that needs to be done more carefully...

**Comment by phil-davis** [comment]
I put this back to developing. Can be progressed next week, depending on priorities.

**Comment by phil-davis** [comment]
Rebased to get up-to-date CI
Assigned myself so I look at this and work out what to do next.

**Comment by kiranparajuli589** [comment]
blocked due to #633

**Comment by kiranparajuli589** [comment]
There are still some failures in `lib/Connection.php`
``` console
php -d zend.enable_gc=0 vendor-bin/phpstan/vendor/bin/phpstan analyse --memory-limit=4G --configuration=./phpstan.neon --no-progress --level=5 appinfo lib
--
34 | ------ ------------------------------------------------------------------
35 | Line   lib/Connection.php
36 | ------ ------------------------------------------------------------------
37 | 560    Parameter #1 $link of method OCA\User_LDAP\ILDAPWrapper::bind()
38 | expects resource, array\|float\|int\|string\|false\|null given.
39 | 569    Parameter #1 $link of method OCA\User_LDAP\ILDAPWrapper::errno()
40 | expects resource, array\|float\|int\|string\|false\|null given.
41 | 569    Parameter #1 $link of method OCA\User_LDAP\ILDAPWrapper::error()
42 | expects resource, array\|float\|int\|string\|false\|null given.
43 | 600    Parameter #1 $link of method OCA\User_LDAP\ILDAPWrapper::bind()
44 | expects resource, array\|float\|int\|string\|false\|null given.
45 | 614    Parameter #1 $link of method OCA\User_LDAP\ILDAPWrapper::errno()
46 | expects resource, array\|float\|int\|string\|false\|null given.
47 | 614    Parameter #1 $link of method OCA\User_LDAP\ILDAPWrapper::error()
48 | expects resource, array\|float\|int\|string\|false\|null given.
49 | ------ ------------------------------------------------------------------
50 |  
51 | [ERROR] Found 6 errors
52 |  
53 | Makefile:127: recipe for target 'test-php-phpstan' failed
54 | make: *** [test-php-phpstan] Error 1
```

**Comment by phil-davis** [comment]
`function installApp missing 1 argument (phpVersion)`
https://drone.owncloud.com/owncloud/user_ldap/3268

**Comment by dpakach** [comment]
> `function installApp missing 1 argument (phpVersion)`
> https://drone.owncloud.com/owncloud/user_ldap/3268

@phil-davis that's the build for different PR.

**Comment by phil-davis** [comment]
> that's the build for different PR.

I pasted in the wrong browser tab!

**Comment by ownclouders** [comment]
:boom: Acceptance tests pipeline <strong>core-cliAll-2-1-master-mysql8.0-php7.3</strong> failed. The build has been cancelled.

https://drone.owncloud.com/owncloud/user_ldap/3618/93/1

**Comment by ownclouders** [comment]
:boom: Acceptance tests pipeline <strong>core-cliAll-2-1-master-mysql8.0-php7.3</strong> failed. The build has been cancelled.

https://drone.owncloud.com/owncloud/user_ldap/3627/93/1

**Comment by ownclouders** [comment]
:boom: Acceptance tests pipeline <strong>core-cliAll-2-1-master-mysql8.0-php7.3</strong> failed. The build has been cancelled.

https://drone.owncloud.com/owncloud/user_ldap/3628/93/1

**Comment by ownclouders** [comment]
:boom: Acceptance tests pipeline <strong>core-cliAll-2-1-master-mysql8.0-php7.3</strong> failed. The build has been cancelled.

https://drone.owncloud.com/owncloud/user_ldap/3631/93/1

**Comment by ownclouders** [comment]
:boom: Acceptance tests pipeline <strong>core-cliAll-2-1-master-mysql8.0-php7.3</strong> failed. The build has been cancelled.

https://drone.owncloud.com/owncloud/user_ldap/3632/93/1

**Comment by phil-davis** [comment]
@dpakach what happened to the acceptance test pipelines?

**Comment by dpakach** [comment]
> @dpakach what happened to the acceptance test pipelines?

working on it

File: lib/Connection.php [hunk 1]
```diff
@@ -364,7 +399,7 @@ private function doSoftValidation() {
 		foreach ($saKeys as $key) {
 			$val = $this->configuration->$key;
 			if (\is_array($val) && \count($val) === 1 && empty($val[0])) {
-				$this->configuration->$key = [];
+				$this->configuration->$key = '';
```
Comment (phil-davis) [inline lib/Connection.php:402]: Comment at line 396 says:
`make sure empty search attributes are saved as simple, empty array`

A simple single empty item is now saved as an empty string.

Does the comment need to be updated?

File: lib/Helper.php [hunk 2]
```diff
@@ -156,10 +156,6 @@ public function haveDisabledConfigurations() {
 		$all = $this->getServerConfigurationPrefixes(false);
 		$active = $this->getServerConfigurationPrefixes(true);
 
-		if (!\is_array($all) || !\is_array($active)) {
-			throw new \Exception('Unexpected Return Value');
-		}
-
```
Comment (phil-davis) [inline lib/Helper.php:162]: This is correct - `getServerConfigurationPrefixes` always returns an array (even if it is empty). So this Exception cannot happen.

File: lib/User/UserEntry.php [hunk 3]
```diff
@@ -266,7 +266,10 @@ public function getEMailAddress() {
 	 */
 	public function getHome() {
 		$path = '';
-		$attr = $this->getAttributeName('homeFolderNamingRule', null);
+		$attr = $this->getAttributeName('homeFolderNamingRule', '');
+		if ($attr === '') {
+			return $attr;
+		}
```
Comment (phil-davis) [inline lib/User/UserEntry.php:272]: Why is this change needed?

File: lib/User_LDAP.php [hunk 4]
```diff
@@ -38,6 +38,7 @@
 use OC\User\NoUserException;
 use OCA\User_LDAP\Exceptions\DoesNotExistOnLDAPException;
 use OCA\User_LDAP\User\Manager;
+use OCA\User_LDAP\User\UserEntry;
```
Comment (phil-davis) [inline lib/User_LDAP.php:41]: Note: this is added so that PHPdoc that refers to it will be parsed correctly.

File: lib/Wizard.php [hunk 5]
```diff
@@ -652,8 +652,8 @@ public function testLoginName($loginName) {
 		\OCP\Util::writeLog('user_ldap', 'Wiz: testLogin '. $loginName . ', filter '. $this->access->getConnection()->ldapLoginFilter, \OCP\Util::DEBUG);
 
 		$users = $this->access->countUsersByLoginName($loginName);
-		if ($this->getLDAP()->errno($cr) !== 0) {
-			throw new \Exception($this->getLDAP()->error($cr));
+		if ($this->getLDAP()->errno($cr) !== 0 && $this->getLDAP()->errno($cr) !== '0') {
```
Comment (phil-davis) [inline lib/Wizard.php:655]: We could save repetition here:
```suggestion
		$errNumber = $this->getLDAP()->errno($cr);
		if ($errNumber !== 0 && $errNumber !== '0') {
```

File: lib/Connection.php (hunk 1 above)
Comment (dpakach) [reply lib/Connection.php:402]: I am not reallly sure why it is that way. Both the properties `'ldapAttributesForUserSearch'` and `'ldapAttributesForGroupSearch'` are supposed to be strings so they should be empty strings logically.

**Review by phil-davis — APPROVED** [review]
